chalk-solve/src/clauses/program_clauses.rs

```diff
@@ -118,6 +118,84 @@ impl<I: Interner> ToProgramClauses<I> for AssociatedTyValue<I> {
     }
 }
 
+impl<I: Interner> ToProgramClauses<I> for OpaqueTyDatum<I> {
+    /// Given `opaque type T<..>: A + B = HiddenTy;`, we generate:
+    ///
+    /// ```notrust
+    /// AliasEq(T<..> = HiddenTy) :- Reveal.
+    /// AliasEq(T<..> = !T<..>).
+    /// Implemented(!T<..>: A).
+    /// Implemented(!T<..>: B).
+    /// Implemented(!T<..>: Send) :- Implemented(HiddenTy: Send). // For all auto traits
+    /// ```
+    /// where `!T<..>` is the placeholder for the unnormalized type `T<..>`.
+    fn to_program_clauses(&self, builder: &mut ClauseBuilder<'_, I>) {
+        let interner = builder.interner();
+        let substitution = builder.substitution_in_scope();
```

Comment:
I suspect this is incorrect. When this is invoked, there won't yet *be* any substitution in scope, you need to (I think) push the binders first.

So move this logic *inside* the call to `push_binders`  below.

In other words, you have `type Foo<P1...Pn> = impl Trait<T1..Tn> = HiddenTy` or whatever, and you want to make some rules like

```
forall<P1..Pn> { // <-- created by the push-binders call
    AliasEq(Foo<P1..Pn> = HiddenTy), // references those P1..Pn
}
```

Reply:
Oh right, all these lines can be moved inside the `push_binders`. :+1: